File: ubag-python/src/ubag/_markdown.py
"""
HTML → Markdown for the honest-fallback content layer.

Tier 1 serves the structured data a site *declares* (typed, verified). But most
pages have prose the owner never encoded as JSON-LD. Rather than *guess* types
for it (a regex that calls "$19.99" a price is a lie an agent will act on), we
convert the readable page content to Markdown and serve it as clearly-labeled
UNSTRUCTURED text. The agent sees "here is the page's readable content," not
"here are verified facts." That honest labeling is the whole point.

Deterministic, stdlib only. Boilerplate (script/style/nav/header/footer/aside/
form) is stripped first so the agent gets content, not chrome.

Note: unlike the JSON-LD path, Markdown output is NOT promised to be byte-
identical across the Python and Node SDKs — HTML-to-text differs by parser. It
is deterministic within each SDK and semantically equivalent across them.
"""
from __future__ import annotations

import re
from html.parser import HTMLParser

_SKIP_TAGS = {
    "head", "title", "script", "style", "noscript", "template", "svg", "nav",
    "header", "footer", "aside", "form", "button", "iframe",
}
_HEADINGS = {"h1", "h2", "h3", "h4", "h5", "h6"}
_BLOCK = {"p", "div", "section", "article", "main", "table", "tr"}


class _MdConverter(HTMLParser):
    def __init__(self) -> None:
        super().__init__(convert_charrefs=True)
        self.parts: list[str] = []
        self._skip = 0
        self._pre = 0
        self._list: list[list] = []  # [tag, counter]
        self._href: str | None = None

    def handle_starttag(self, tag: str, attrs: list) -> None:
        if tag in _SKIP_TAGS:
            self._skip += 1
            return
        if self._skip:
            return
        a = {k.lower(): (v or "") for k, v in attrs}
        if tag in _HEADINGS:
            self.parts.append("\n\n" + "#" * int(tag[1]) + " ")
        elif tag in _BLOCK:
            self.parts.append("\n\n")
        elif tag == "br":
            self.parts.append("\n")
        elif tag == "hr":
            self.parts.append("\n\n---\n\n")
        elif tag in ("ul", "ol"):
            self._list.append([tag, 0])
            self.parts.append("\n")
        elif tag == "li":
            indent = "  " * max(0, len(self._list) - 1)
            if self._list and self._list[-1][0] == "ol":
                self._list[-1][1] += 1
                self.parts.append(f"\n{indent}{self._list[-1][1]}. ")
            else:
                self.parts.append(f"\n{indent}- ")
        elif tag in ("strong", "b"):
            self.parts.append("**")
        elif tag in ("em", "i"):
            self.parts.append("*")
        elif tag == "code" and not self._pre:
            self.parts.append("`")
        elif tag == "pre":
            self._pre += 1
            self.parts.append("\n\n```\n")
        elif tag == "blockquote":
            self.parts.append("\n\n> ")
        elif tag == "a":
            self._href = a.get("href") or None
            self.parts.append("[")

    def handle_endtag(self, tag: str) -> None:
        if tag in _SKIP_TAGS:
            self._skip = max(0, self._skip - 1)
            return
        if self._skip:
            return
        if tag in ("strong", "b"):
            self.parts.append("**")
        elif tag in ("em", "i"):
            self.parts.append("*")
        elif tag == "code" and not self._pre:
            self.parts.append("`")
        elif tag == "pre":
            self._pre = max(0, self._pre - 1)
            self.parts.append("\n```\n\n")
        elif tag in ("ul", "ol"):
            if self._list:
                self._list.pop()
            self.parts.append("\n")
        elif tag == "a":
            href, self._href = self._href, None
            self.parts.append(f"]({href})" if href else "]")
        elif tag in _HEADINGS or tag in _BLOCK:
            self.parts.append("\n")

    def handle_data(self, data: str) -> None:
        if self._skip:
            return
        if self._pre:
            self.parts.append(data)
        else:
            self.parts.append(re.sub(r"\s+", " ", data))


def _strip_boilerplate(html: str) -> str:
    for tag in _SKIP_TAGS:
        html = re.sub(
            rf"<{tag}\b[^>]*>.*?</{tag}>", " ", html, flags=re.IGNORECASE | re.DOTALL
        )
    return html


def _normalize(md: str) -> str:
    md = re.sub(r"[ \t]+\n", "\n", md)     # strip trailing spaces on lines
    md = re.sub(r"\n{3,}", "\n\n", md)     # collapse runs of blank lines
    md = re.sub(r"[ \t]{2,}", " ", md)     # collapse internal runs of spaces
    return md.strip()


def html_to_markdown(html: str, max_chars: int | None = None) -> str:
    """Convert an HTML document to readable Markdown. Boilerplate is removed;
    output is deterministic. Truncated to `max_chars` with a marker if set.
    """
    conv = _MdConverter()
    try:
        conv.feed(_strip_boilerplate(html or ""))
    except Exception:
        pass
    md = _normalize("".join(conv.parts))
    if max_chars and len(md) > max_chars:
        md = md[:max_chars].rstrip() + "\n\n… [truncated]"
    return md

File: ubag-node/src/markdown.js
'use strict';

/**
 * HTML → Markdown for the honest-fallback content layer.
 * Mirrors ubag-python/src/ubag/_markdown.py.
 *
 * Serves readable page content as clearly-labeled UNSTRUCTURED Markdown instead
 * of guessing types for prose the owner never encoded. Boilerplate is stripped
 * first. Deterministic within this SDK; not promised byte-identical to Python
 * (HTML-to-text differs by parser), but semantically equivalent.
 */

const SKIP_TAGS = new Set([
  'head', 'title', 'script', 'style', 'noscript', 'template', 'svg', 'nav',
  'header', 'footer', 'aside', 'form', 'button', 'iframe',
]);
const HEADINGS = new Set(['h1', 'h2', 'h3', 'h4', 'h5', 'h6']);
const BLOCK = new Set(['p', 'div', 'section', 'article', 'main', 'table', 'tr']);

function decodeEntities(s) {
  return String(s)
    .replace(/&amp;/g, '&').replace(/&lt;/g, '<').replace(/&gt;/g, '>')
    .replace(/&quot;/g, '"').replace(/&#0*39;/g, "'").replace(/&apos;/g, "'")
    .replace(/&nbsp;/g, ' ');
}

function hrefOf(attrs) {
  const m = /\bhref\s*=\s*("([^"]*)"|'([^']*)'|([^\s"'>]+))/i.exec(attrs);
  if (!m) return null;
  return (m[2] !== undefined ? m[2] : m[3] !== undefined ? m[3] : m[4] || '').trim() || null;
}

function stripBoilerplate(html) {
  for (const tag of SKIP_TAGS) {
    html = html.replace(new RegExp(`<${tag}\\b[^>]*>[\\s\\S]*?<\\/${tag}>`, 'gi'), ' ');
  }
  return html;
}

function normalize(md, maxChars) {
  md = md
    .replace(/[ \t]+\n/g, '\n')
    .replace(/\n{3,}/g, '\n\n')
    .replace(/[ \t]{2,}/g, ' ')
    .trim();
  if (maxChars && md.length > maxChars) {
    md = md.slice(0, maxChars).replace(/\s+$/, '') + '\n\n… [truncated]';
  }
  return md;
}

function htmlToMarkdown(html, maxChars = null) {
  html = stripBoilerplate(String(html || ''));
  const parts = [];
  const st = { skip: 0, pre: 0, list: [], href: null };
  const tagRe = /<(\/?)([a-zA-Z0-9]+)((?:[^>"']|"[^"]*"|'[^']*')*)\/?>/g;
  let last = 0;
  let m;
  while ((m = tagRe.exec(html))) {
    if (m.index > last) handleText(parts, st, html.slice(last, m.index));
    last = tagRe.lastIndex;
    const tag = m[2].toLowerCase();
    if (m[1] === '/') handleEnd(parts, st, tag);
    else handleStart(parts, st, tag, m[3]);
  }
  if (last < html.length) handleText(parts, st, html.slice(last));
  return normalize(parts.join(''), maxChars);
}

function handleStart(parts, st, tag, attrs) {
  if (SKIP_TAGS.has(tag)) {
    st.skip += 1;
    return;
  }
  if (st.skip) return;
  if (HEADINGS.has(tag)) parts.push('\n\n' + '#'.repeat(parseInt(tag[1], 10)) + ' ');
  else if (BLOCK.has(tag)) parts.push('\n\n');
  else if (tag === 'br') parts.push('\n');
  else if (tag === 'hr') parts.push('\n\n---\n\n');
  else if (tag === 'ul' || tag === 'ol') {
    st.list.push([tag, 0]);
    parts.push('\n');
  } else if (tag === 'li') {
    const indent = '  '.repeat(Math.max(0, st.list.length - 1));
    const top = st.list[st.list.length - 1];
    if (top && top[0] === 'ol') {
      top[1] += 1;
      parts.push(`\n${indent}${top[1]}. `);
    } else parts.push(`\n${indent}- `);
  } else if (tag === 'strong' || tag === 'b') parts.push('**');
  else if (tag === 'em' || tag === 'i') parts.push('*');
  else if (tag === 'code' && !st.pre) parts.push('`');
  else if (tag === 'pre') {
    st.pre += 1;
    parts.push('\n\n```\n');
  } else if (tag === 'blockquote') parts.push('\n\n> ');
  else if (tag === 'a') {
    st.href = hrefOf(attrs);
    parts.push('[');
  }
}

function handleEnd(parts, st, tag) {
  if (SKIP_TAGS.has(tag)) {
    st.skip = Math.max(0, st.skip - 1);
    return;
  }
  if (st.skip) return;
  if (tag === 'strong' || tag === 'b') parts.push('**');
  else if (tag === 'em' || tag === 'i') parts.push('*');
  else if (tag === 'code' && !st.pre) parts.push('`');
  else if (tag === 'pre') {
    st.pre = Math.max(0, st.pre - 1);
    parts.push('\n```\n\n');
  } else if (tag === 'ul' || tag === 'ol') {
    st.list.pop();
    parts.push('\n');
  } else if (tag === 'a') {
    const href = st.href;
    st.href = null;
    parts.push(href ? `](${href})` : ']');
  } else if (HEADINGS.has(tag) || BLOCK.has(tag)) parts.push('\n');
}

function handleText(parts, st, text) {
  if (st.skip) return;
  if (st.pre) parts.push(decodeEntities(text));
  else parts.push(decodeEntities(text).replace(/\s+/g, ' '));
}

module.exports = { htmlToMarkdown };
